File: src/fr/cph/stock/android/activity/OverallActivity.java
/**
 * Copyright 2013 Carl-Philipp Harmant
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.stock.android.activity;

import android.app.ActionBar.LayoutParams;
import android.app.Activity;
import android.app.Dialog;
import android.app.ListActivity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.Spinner;
import android.widget.TextView;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import fr.cph.stock.android.R;
import fr.cph.stock.android.StockTrackerApp;
import fr.cph.stock.android.adapter.ShareValueAdapter;
import fr.cph.stock.android.entity.Account;
import fr.cph.stock.android.entity.Portfolio;
import fr.cph.stock.android.entity.ShareValue;
import fr.cph.stock.android.enumtype.UrlType;
import fr.cph.stock.android.task.MainTask;

import static fr.cph.stock.android.Constants.PORTFOLIO;

public class OverallActivity extends ListActivity implements IStockTrackerActivity {

	private static final String TAG = "OverallActivity";

	private MenuItem refreshItem;
	private List<ShareValue> shareValues;
	private ShareValueAdapter ada;
	private TextView errorView;
	private Portfolio portfolio;

	@Override
	protected void onCreate(final Bundle bundle) {
		super.onCreate(bundle);
		setContentView(R.layout.overall_activity);
		portfolio = bundle != null
				? bundle.getParcelable(PORTFOLIO)
				: getIntent().getParcelableExtra(PORTFOLIO);
		errorView = findViewById(R.id.errorMessage);
		shareValues = portfolio.getShareValues();
		ada = new ShareValueAdapter(shareValues, getApplicationContext());
		setListAdapter(ada);
	}

	@Override
	public boolean onCreateOptionsMenu(final Menu menu) {
		// Inflate the menu; this adds items to the action bar if it is present.
		getMenuInflater().inflate(R.menu.overall, menu);
		refreshItem = menu.getItem(2);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(final MenuItem item) {
		MainTask mainTask;
		switch (item.getItemId()) {
			case R.id.action_logout:
				mainTask = new MainTask(this, UrlType.LOGOUT, null);
				mainTask.execute((Void) null);
				return true;
			case R.id.refresh:
				item.setActionView(R.layout.progressbar);
				item.expandActionView();
				mainTask = new MainTask(this, UrlType.RELOAD, null);
				mainTask.execute((Void) null);
				return true;
			case R.id.action_update:
				showPanelUpdateHistory();
				return true;
			case android.R.id.home:
				finish();
				return true;
			default:
				return super.onOptionsItemSelected(item);
		}
	}

	private void showPanelUpdateHistory() {
		final Dialog alert = new Dialog(this);
		alert.setTitle("Update history");
		alert.setContentView(R.layout.history_dialog);

		final Spinner checked = alert.findViewById(R.id.accountList);
		List<String> list = new ArrayList<>();
		for (Account acc : portfolio.getAccounts()) {
			list.add(acc.getName());
		}
		ArrayAdapter<String> dataAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, list);
		dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
		checked.setAdapter(dataAdapter);

		Button dialogButton = alert.findViewById(R.id.dialogButtonOK);
		dialogButton.setOnClickListener(v -> {
			refreshItem.setActionView(R.layout.progressbar);
			refreshItem.expandActionView();
			Account account = portfolio.getAccounts().get(checked.getSelectedItemPosition());
			EditText liquidityView = alert.findViewById(R.id.liquidityMov);
			EditText yieldView = alert.findViewById(R.id.yield);
			EditText buyView = alert.findViewById(R.id.buy);
			EditText sellView = alert.findViewById(R.id.sell);
			EditText taxeView = alert.findViewById(R.id.taxe);
			EditText commentaryView = alert.findViewById(R.id.commentaryEditText);

			String params = "?accountId=" + account.getId() + "&liquidity=" + liquidityView.getText() + "&yield="
					+ yieldView.getText() + "&buy=" + buyView.getText() + "&sell=" + sellView.getText() + "&taxe="
					+ taxeView.getText() + "&commentary=" + commentaryView.getText().toString().replaceAll(" ", "%20");
			MainTask mainTask = new MainTask(OverallActivity.this, UrlType.UPDATEHISTORY, params);
			mainTask.execute((Void) null);
			alert.dismiss();
		});
		dialogButton = alert.findViewById(R.id.dialogButtonCancel);
		dialogButton.setOnClickListener(v -> alert.dismiss());
		alert.show();
	}

	@Override
	public void reloadData(final Portfolio portfolio) {
		shareValues.clear();
		shareValues.addAll(portfolio.getShareValues());
		ada.notifyDataSetChanged();
		refreshItem.collapseActionView();
		refreshItem.setActionView(null);
		Intent resultIntent = new Intent();
		resultIntent.putExtra(PORTFOLIO, portfolio);
		setResult(Activity.RESULT_OK, resultIntent);
		StockTrackerApp app = (StockTrackerApp) getApplication();
		app.toast();
	}

	@Override
	public void displayError(final JSONObject json) {
		Log.d(TAG, json.toString());
		boolean sessionError = ((StockTrackerApp) getApplication()).isSessionError(json);
		if (sessionError) {
			((StockTrackerApp) getApplication()).loadErrorActivity(this, json);
		} else {
			errorView.setText(json.optString("error"));
			RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT);
			params.addRule(RelativeLayout.BELOW, errorView.getId());
			ListView listView = findViewById(android.R.id.list);
			listView.setLayoutParams(params);
			refreshItem.collapseActionView();
			refreshItem.setActionView(null);
		}
	}

	@Override
	public void logOut() {
		((StockTrackerApp) getApplication()).logOut(this);
	}
}

File: src/fr/cph/stock/android/activity/OverallActivity.kt
/**
 * Copyright 2017 Carl-Philipp Harmant
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.stock.android.activity

import android.app.ActionBar.LayoutParams
import android.app.Activity
import android.app.Dialog
import android.app.ListActivity
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.*
import fr.cph.stock.android.Constants
import fr.cph.stock.android.R
import fr.cph.stock.android.StockTrackerApp
import fr.cph.stock.android.adapter.ShareValueAdapter
import fr.cph.stock.android.domain.Portfolio
import fr.cph.stock.android.domain.ShareValue
import fr.cph.stock.android.domain.UrlType
import fr.cph.stock.android.task.MainTask
import org.json.JSONObject
import java.util.*

class OverallActivity : ListActivity(), StockTrackerActivity {

    private lateinit var refreshItem: MenuItem
    private lateinit var shareValues: MutableList<ShareValue>
    private lateinit var ada: ShareValueAdapter
    private lateinit var errorView: TextView
    private lateinit var portfolio: Portfolio

    override fun onCreate(bundle: Bundle?) {
        super.onCreate(bundle)
        setContentView(R.layout.overall_activity)
        portfolio = if (bundle != null)
            bundle.getParcelable(Constants.PORTFOLIO)
        else
            intent.getParcelableExtra(Constants.PORTFOLIO)
        errorView = findViewById(R.id.errorMessage)
        shareValues = portfolio.shareValues.toMutableList()
        ada = ShareValueAdapter(shareValues, applicationContext)
        listAdapter = ada
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.overall, menu)
        refreshItem = menu.getItem(2)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.action_logout -> consume { MainTask(this, UrlType.LOGOUT, emptyMap()).execute(null as Void?) }
        R.id.refresh -> consume {
            item.setActionView(R.layout.progressbar)
            item.expandActionView()
            MainTask(this, UrlType.RELOAD, emptyMap()).execute(null as Void?)
        }
        R.id.action_update -> consume { showPanelUpdateHistory() }
        android.R.id.home -> consume { finish() }
        else -> super.onOptionsItemSelected(item)
    }

    private fun showPanelUpdateHistory() {
        val alert = Dialog(this)
        alert.setTitle("Update history")
        alert.setContentView(R.layout.history_dialog)

        val checked = alert.findViewById<Spinner>(R.id.accountList)
        val list = portfolio.accounts.map { it.name }
        val dataAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, list)
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        checked.adapter = dataAdapter

        var dialogButton = alert.findViewById<Button>(R.id.dialogButtonOK)
        dialogButton.setOnClickListener { _ ->
            refreshItem.setActionView(R.layout.progressbar)
            refreshItem.expandActionView()
            val account = portfolio.accounts[checked.selectedItemPosition]
            val liquidityView = alert.findViewById<EditText>(R.id.liquidityMov)
            val yieldView = alert.findViewById<EditText>(R.id.yield)
            val buyView = alert.findViewById<EditText>(R.id.buy)
            val sellView = alert.findViewById<EditText>(R.id.sell)
            val taxeView = alert.findViewById<EditText>(R.id.taxe)
            val commentaryView = alert.findViewById<EditText>(R.id.commentaryEditText)

            val params = hashMapOf(
                    "accountId" to account.id,
                    "liquidity" to liquidityView.text.toString(),
                    "yield" to yieldView.text.toString(),
                    "buy" to buyView.text.toString(),
                    "sell" to sellView.text.toString(),
                    "taxe" to taxeView.text.toString(),
                    "commentary" to commentaryView.text.toString().replace(" ".toRegex(), "%20")
            )
            val mainTask = MainTask(this@OverallActivity, UrlType.UPDATEHISTORY, params)
            mainTask.execute(null as Void?)
            alert.dismiss()
        }
        dialogButton = alert.findViewById(R.id.dialogButtonCancel)
        dialogButton.setOnClickListener { _ -> alert.dismiss() }
        alert.show()
    }

    override fun update(portfolio: Portfolio) {
        shareValues.clear()
        shareValues.addAll(portfolio.shareValues)
        ada.notifyDataSetChanged()
        refreshItem.collapseActionView()
        refreshItem.actionView = null
        val resultIntent = Intent()
        resultIntent.putExtra(Constants.PORTFOLIO, portfolio)
        setResult(Activity.RESULT_OK, resultIntent)
        val app = application as StockTrackerApp
        app.toast()
    }

    override fun displayError(json: JSONObject) {
        Log.d(TAG, json.toString())
        val sessionError = (application as StockTrackerApp).isSessionError(json)
        if (sessionError) {
            (application as StockTrackerApp).loadErrorActivity(this, json)
        } else {
            errorView.text = json.optString("error")
            val params = RelativeLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
            params.addRule(RelativeLayout.BELOW, errorView.id)
            val listView = findViewById<ListView>(android.R.id.list)
            listView.layoutParams = params
            refreshItem.collapseActionView()
            refreshItem.actionView = null
        }
    }

    override fun logOut() {
        (application as StockTrackerApp).logOut(this)
    }

    companion object {
        private val TAG = "OverallActivity"
    }
}
